pkg/code-generator/schemagen/protoc.go

```diff
@@ -0,0 +1,211 @@
+
+			// Spec validation schema
+			specJsonSchema, err := p.getJsonSchema(schemaKey, schemaValue)
+			if err != nil {
+				return err
+			}
+
+			jsonSchemasByGVK[schemaGVK] = specJsonSchema
+		}
+		// Continue traversing the output directory
+		return nil
+	})
+
+	return jsonSchemasByGVK, err
+}
+
+func (p *protocGenerator) readOpenApiDocumentFromFile(file string) (*openapi3.Swagger, error) {
+	var openApiDocument *openapi3.Swagger
+	bytes, err := ioutil.ReadFile(file)
+	if err != nil {
+		return nil, errors.Wrapf(err, "reading file")
+	}
+	if err := yaml.Unmarshal(bytes, &openApiDocument); err != nil {
+		return nil, errors.Wrapf(err, "unmarshalling tmp file as schemas")
+	}
+	return openApiDocument, nil
+}
+
+func (p *protocGenerator) getGVKForSchemaKey(project *model.Project, schemaKey string) schema.GroupVersionKind {
+	// The generated keys look like testing.solo.io.MockResource
+	// The kind is the `MockResource` portion
+	ss := strings.Split(schemaKey, ".")
+	kind := ss[len(ss)-1]
+
+	projectGV := model.GetGVForProject(project)
+
+	return schema.GroupVersionKind{
+		Group:   projectGV.Group,
+		Version: projectGV.Version,
+		Kind:    kind,
+	}
+}
+
+func (p *protocGenerator) getJsonSchema(schemaKey string, schema *openapi3.SchemaRef) (*v1beta1.JSONSchemaProps, error) {
+	if schema == nil {
+		return nil, eris.Errorf("no open api schema for %s", schemaKey)
+	}
+
+	oApiJson, err := schema.MarshalJSON()
+	if err != nil {
+		return nil, eris.Errorf("Cannot marshal OpenAPI schema for %v: %v", schemaKey, err)
+	}
+
+	var obj map[string]interface{}
+	if err = json.Unmarshal(oApiJson, &obj); err != nil {
+		return nil, err
+	}
+
+	p.removeProtoAnyValidation(obj)
+
+	bytes, err := json.Marshal(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	jsonSchema := &v1beta1.JSONSchemaProps{}
+	if err = json.Unmarshal(bytes, jsonSchema); err != nil {
+		return nil, eris.Errorf("Cannot unmarshal raw OpenAPI schema to JSONSchemaProps for %v: %v", schemaKey, err)
+	}
+
+	return jsonSchema, nil
+}
+
+// prevent k8s from validating proto.Any fields (since it's unstructured)
+func (p *protocGenerator) removeProtoAnyValidation(d map[string]interface{}) {
+	for _, v := range d {
+		values, ok := v.(map[string]interface{})
+		if !ok {
+			continue
+		}
+		desc, ok := values["properties"]
+		properties, isObj := desc.(map[string]interface{})
+		// detect proto.Any field from presence of "type_url" as field under "properties"
+		if !ok || !isObj || properties["type_url"] == nil {
+			p.removeProtoAnyValidation(values)
+			continue
+		}
+		// remove "properties" value
+		delete(values, "properties")
+		// remove "required" value
+		delete(values, "required")
+		// x-kubernetes-preserve-unknown-fields allows for unknown fields from a particular node
+		// see https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema
+		values["x-kubernetes-preserve-unknown-fields"] = true
+	}
+}
```

Comment:
move shared logic to a shared package?

Reply:
Happy to! They're slightly different implementations (one is looking for type_url, the other for @type). Though that can easily be handled